base/floatfuncs.jl

```diff
@@ -276,6 +276,9 @@ significantly more expensive than `x*y+z`. `fma` is used to improve accuracy in
 algorithms. See [`muladd`](@ref).
 """
 function fma end
+function fma_emulated(a::Float16, b::Float16, c::Float16)
+    Float16(muladd(Float32(a), Float32(b), Float32(c))) #don't use fma if the hardware doesn't have it.
```

Comment:
I think this can be simplified to
```suggestion
    Float16(muladd(a, b, c)) #don't use fma if the hardware doesn't have it.
```
LLVM would automatically do the demotion to `float` as necessary nowadays.

Reply:
Ironically, on aarch64 with fp16 extension `muladd` is better than `fma` on Float16 because it doesn't force the Float16 -> Float32 -> Float16 dance:
```
julia> code_native(muladd, NTuple{3,Float16}; debuginfo=:none)
```
```asm
        .text
        .file   "muladd"
        .globl  julia_muladd_1256               // -- Begin function julia_muladd_1256
        .p2align        4
        .type   julia_muladd_1256,@function
julia_muladd_1256:                      // @julia_muladd_1256
; Function Signature: muladd(Float16, Float16, Float16)
// %bb.0:                               // %top
        //DEBUG_VALUE: muladd:x <- $h0
        //DEBUG_VALUE: muladd:x <- $h0
        //DEBUG_VALUE: muladd:y <- $h1
        //DEBUG_VALUE: muladd:y <- $h1
        //DEBUG_VALUE: muladd:z <- $h2
        //DEBUG_VALUE: muladd:z <- $h2
        stp     x29, x30, [sp, #-16]!           // 16-byte Folded Spill
        mov     x29, sp
        fmadd   h0, h0, h1, h2
        ldp     x29, x30, [sp], #16             // 16-byte Folded Reload
        ret
.Lfunc_end0:
        .size   julia_muladd_1256, .Lfunc_end0-julia_muladd_1256
                                        // -- End function
        .type   ".L+Core.Float16#1258",@object  // @"+Core.Float16#1258"
        .section        .rodata,"a",@progbits
        .p2align        3, 0x0
".L+Core.Float16#1258":
        .xword  ".L+Core.Float16#1258.jit"
        .size   ".L+Core.Float16#1258", 8

.set ".L+Core.Float16#1258.jit", 281472349230944
        .size   ".L+Core.Float16#1258.jit", 8
        .section        ".note.GNU-stack","",@progbits
```
```julia-repl
julia> code_native(fma, NTuple{3,Float16}; debuginfo=:none)
```
```asm
        .text
        .file   "fma"
        .globl  julia_fma_1259                  // -- Begin function julia_fma_1259
        .p2align        4
        .type   julia_fma_1259,@function
julia_fma_1259:                         // @julia_fma_1259
; Function Signature: fma(Float16, Float16, Float16)
// %bb.0:                               // %top
        //DEBUG_VALUE: fma:a <- $h0
        //DEBUG_VALUE: fma:a <- $h0
        //DEBUG_VALUE: fma:b <- $h1
        //DEBUG_VALUE: fma:b <- $h1
        //DEBUG_VALUE: fma:c <- $h2
        //DEBUG_VALUE: fma:c <- $h2
        stp     x29, x30, [sp, #-16]!           // 16-byte Folded Spill
        fcvt    s0, h0
        fcvt    s1, h1
        mov     x29, sp
        fcvt    s2, h2
        fmadd   s0, s0, s1, s2
        fcvt    h0, s0
        ldp     x29, x30, [sp], #16             // 16-byte Folded Reload
        ret
.Lfunc_end0:
        .size   julia_fma_1259, .Lfunc_end0-julia_fma_1259
                                        // -- End function
        .type   ".L+Core.Float16#1261",@object  // @"+Core.Float16#1261"
        .section        .rodata,"a",@progbits
        .p2align        3, 0x0
".L+Core.Float16#1261":
        .xword  ".L+Core.Float16#1261.jit"
        .size   ".L+Core.Float16#1261", 8

.set ".L+Core.Float16#1261.jit", 281472349230944
        .size   ".L+Core.Float16#1261.jit", 8
        .section        ".note.GNU-stack","",@progbits
```

Reply:
no. `muladd` doesn't guarantee the accuracy of `fma` requires

Reply:
I will also point out that pure f16 fma is not super useful as an operation. Most of the accelerators will do fp16 multiply with an f32 accumulator (and then potentially round back to f16 at the end).

Reply:
sure, but that's not what `Base.fma` does.

Reply:
That’s not really true of fp16, on aarch64 it’s a true type which supports everything (with twice the throughput on SIMD), bf16 is that though